spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -172,171 +149,69 @@ public RedisLockRegistry(RedisConnectionFactory connectionFactory, String regist
-			getHardThreadLocks().remove(lock);
-		}
-
-		if (this.useWeakReferences) {
-			getWeakThreadLocks().add(lock);
-		}
-
-		//clean up
-		if (this.hardThreadLocks.get() != null && this.hardThreadLocks.get().isEmpty()) {
-			this.hardThreadLocks.remove();
-		}
 	}
 
 	@Override
 	public Lock obtain(Object lockKey) {
 		Assert.isInstanceOf(String.class, lockKey);
-
-		//try to find the lock within hard references
-		RedisLock lock = findLock(this.hardThreadLocks.get(), lockKey);
-
-		/*
-		 * If the lock is locked, check that it matches what's in the store.
-		 * If it doesn't, the lock must have expired.
-		 */
-		if (lock != null && lock.thread != null) {
-			RedisLock lockInStore = this.redisTemplate.boundValueOps(this.registryKey + ":" + lockKey).get();
-			if (lockInStore == null || !lock.equals(lockInStore)) {
-				try {
-					lock.unlock();
-				}
-				catch (Exception e) {
-					if (logger.isWarnEnabled()) {
-						logger.warn("Lock was released due to expiration. A new one will be obtained...", e);
-					}
-				}
-				if (this.hardThreadLocks.get() != null) {
-					this.hardThreadLocks.get().remove(lock);
-				}
-				if (this.weakThreadLocks.get() != null) {
-					this.weakThreadLocks.get().remove(lock);
-				}
-				lock = null;
+		String path = (String) lockKey;
+		RedisLock lock;
+		synchronized (this.locks) {
+			lock = this.locks.get(path);
+			if (lock == null) {
+				lock = new RedisLock(path);
+				this.locks.put(path, lock);
 			}
 		}
+		return lock;
+	}
 
-		if (lock == null) {
-			//try to find the lock within weak references
-			lock = findLock(this.weakThreadLocks.get(), lockKey);
-
-			if (lock == null) {
-				lock = new RedisLock((String) lockKey);
-
-				if (this.useWeakReferences) {
-					getWeakThreadLocks().add(lock);
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		synchronized (this.locks) {
+			Iterator<Map.Entry<String, RedisLock>> iterator = this.locks.entrySet().iterator();
+			long now = System.currentTimeMillis();
+			while (iterator.hasNext()) {
+				Map.Entry<String, RedisLock> entry = iterator.next();
+				RedisLock lock = entry.getValue();
+				if (now - lock.getLockedAt() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
 				}
 			}
 		}
-
-		return lock;
 	}
 
 	public Collection<Lock> listLocks() {
-		return this.redisTemplate.execute(new RedisCallback<Collection<Lock>>() {
-
-			@Override
-			public Collection<Lock> doInRedis(RedisConnection connection) throws DataAccessException {
-				Set<byte[]> keys = connection.keys((RedisLockRegistry.this.registryKey + ":*").getBytes());
-				ArrayList<Lock> list = new ArrayList<Lock>(keys.size());
-				if (keys.size() > 0) {
-					List<byte[]> locks = connection.mGet(keys.toArray(new byte[keys.size()][]));
-					for (byte[] lock : locks) {
-						list.add(RedisLockRegistry.this.lockSerializer.deserialize(lock));
-					}
-				}
-				return list;
-			}
-		});
+		return Collections.<Lock>unmodifiableCollection(this.locks.values());
```

Comment:
We should store additional fields in the hash that represents the lock (currently there's only `clientId`), and use [`HGETALL`](https://redis.io/commands/hgetall) to load the entire hash and reconstruct the `RedisLock`.

JSON serialization isn't suitable here as we need to consider the Lua script for obtaining lock - we don't want to parse `clientId` from a JSON in there :smile:

> Although this API is pretty arbitrary and similar in the `JdbcLockRegisrty` returns only local instances, and even any other impls don't provide such an API, I think that should not be so critical to be consistent with existing functionality and make this method to return all the locks from the Redis.

Can you clarify the _similar in the `JdbcLockRegisrty` returns only local instances_ part a bit? AFAICT `JdbcLockRegistry` doesn't offer functionality similar to `RedisLockRegistry#listLocks`.

Also could we consider providing this in `4.3.x`, to maintain backwards compatibility, and deprecate `RedisLockRegistry#listLocks` and remove it in `5.0`? It's not a part of `LockRegistry` contract and as you say _the API is pretty arbitrary_. IMO it's unfortunate that such method forces us to make our storage more complicated than they would've been otherwise.